tests/extras/datasets/hdf5/test_h5py_dataset.py

```diff
@@ -0,0 +1,259 @@
+from pathlib import Path, PurePosixPath
+from typing import Union
+
+import h5py
+import numpy as np
+import pytest
+from fsspec.implementations.http import HTTPFileSystem
+from fsspec.implementations.local import LocalFileSystem
+from gcsfs import GCSFileSystem
+from s3fs.core import S3FileSystem
+
+from kedro.extras.datasets.hdf5.h5py_dataset import H5pyDataSet
+from kedro.io import DataSetError
+from kedro.io.core import PROTOCOL_DELIMITER, Version
+
+
+@pytest.fixture
+def filepath_hdf(tmp_path):
+    return (tmp_path / "test.h5").as_posix()
+
+
+@pytest.fixture
+def hdf_data_set(filepath_hdf, load_args, save_args, mocker, fs_args):
+    H5pyDataSet._lock = mocker.MagicMock()
+    return H5pyDataSet(
+        filepath=filepath_hdf,
+        load_args=load_args,
+        save_args=save_args,
+        fs_args=fs_args,
+    )
+
+
+@pytest.fixture
+def versioned_hdf_data_set(filepath_hdf, load_version, save_version):
+    return H5pyDataSet(
+        filepath=filepath_hdf, version=Version(load_version, save_version)
+    )
+
+
+@pytest.fixture
+def dummy_filepath_hdf(tmp_path):
+    return (tmp_path / "dummy.h5").as_posix()
+
+
+@pytest.fixture
+def dummy_h5f(dummy_filepath_hdf):
+    h5f = h5py.File(dummy_filepath_hdf, "w")
+    group = h5f.create_group("foo")
+    group.create_dataset("a", data=np.arange(4).reshape((2, 2)))
+    h5f.create_dataset("b", data=np.array([1, 2]))
+    return h5f
+
+
+def assert_h5py_equal(
+    expected: Union[h5py.File, h5py.Dataset], received: Union[h5py.File, h5py.Dataset]
+) -> bool:
+
+    try:
+        if isinstance(expected, h5py.Dataset) or isinstance(received, h5py.Dataset):
+            assert isinstance(expected, h5py.Dataset) and isinstance(
+                received, h5py.Dataset
+            )
+            # Indexing a Dataset with () returns all its values
+            # which in this case are assumed to be numpy arrays
+            np.testing.assert_equal(expected[()], received[()])
```

Comment:
Can we remove this function and simplified this as

```
assert isinstance(expected, h5py.Dataset) and isinstance(
    received, h5py.Dataset
)
np.testing.assert_equal(expected[()], received[()])
```

The function assumes both object need to be h5py.DataSet otherwise it will fail, so there is no need to do type checking.
In addition the boolean value isn't used in the following test so this seems not necessary.